src/MSBuild/XMake.cs

```diff
@@ -2765,62 +2765,53 @@ DirectoryGetFiles getFiles
                 // Get all files in the current directory that have a proj-like extension
                 string[] potentialProjectFiles = getFiles(projectDirectory ?? ".", "*.*proj");
-                // Get all files in the current directory that have a sln extension
-                string[] potentialSolutionFiles = getFiles(projectDirectory ?? ".", "*.sln");
-
-                List<string> extensionsToIgnore = new List<string>();
-                if (projectsExtensionsToIgnore != null)
-                {
-                    extensionsToIgnore.AddRange(projectsExtensionsToIgnore);
-                }
-
                 if (potentialProjectFiles != null)
                 {
+                    List<string> actualProjectFiles = new List<string>(potentialProjectFiles.Length);
                     foreach (string s in potentialProjectFiles)
                     {
-                        if (s.EndsWith("~", StringComparison.CurrentCultureIgnoreCase))
+                        if (!extensionsToIgnore.Contains(Path.GetExtension(s)) && !s.EndsWith("~", StringComparison.CurrentCultureIgnoreCase))
                         {
-                            extensionsToIgnore.Add(Path.GetExtension(s));
+                            actualProjectFiles.Add(s);
                         }
                     }
+                    potentialProjectFiles = actualProjectFiles.ToArray();
                 }
 
+                // Get all files in the current directory that have a sln extension
+                string[] potentialSolutionFiles = getFiles(projectDirectory ?? ".", "*.sln");
+                string[] solutionFilterFiles = Array.Empty<string>();
                 if (potentialSolutionFiles != null)
                 {
+                    List<string> actualSolutionFiles = new List<string>(potentialSolutionFiles.Length);
+                    List<string> actualSolutionFilterFiles = new List<string>(potentialSolutionFiles.Length);
                     foreach (string s in potentialSolutionFiles)
                     {
-                        if (!FileUtilities.IsSolutionFilename(s))
+                        if (!extensionsToIgnore.Contains(Path.GetExtension(s)))
                         {
-                            extensionsToIgnore.Add(Path.GetExtension(s));
+                            if (FileUtilities.IsSolutionFilterFilename(s))
+                            {
+                                actualSolutionFilterFiles.Add(s);
+                            }
+                            else if (FileUtilities.IsSolutionFilename(s))
+                            {
+                                actualSolutionFiles.Add(s);
+                            }
                         }
                     }
+                    potentialSolutionFiles = actualSolutionFiles.ToArray();
+                    solutionFilterFiles = actualSolutionFiles.ToArray();
```

Comment:
Do you need to array-ify these? Since they were accumulated as a list and are short-lived local variables (so they can be GCed at return), I don't think there's a big benefit.